wallet-new/test/unit/UTxO/BlockGen.hs

```diff
@@ -219,3 +236,175 @@ zipFees
+      --   Only the non-integer part of this value is considered.
+    , _treeGenGlobalCtxPruneLikelihood             :: !Double
+      -- ^ Likelihood of the non-principal branch to be terminated at any
+      --   block.
+    , _treeGenGlobalCtxMaxHeight                   :: !Int
+      -- ^ Maximum height of the tree. The principal branch will be grown to
+      --   this height. In order to preserve the property that we only switch
+      --   to longer forks, if any other branch grows to this height, it will
+      --   be terminated at (maxHeight - 1).
+    , _treeGenGlobalCtxSharedTransactionLikelihood :: !Double
+      -- ^ The likelihood that a given transaction will be shared between two
+      --   blocks with the same parent.
+    , _treeGenGlobalCtxBootTransaction             :: Transaction h Addr
+      -- ^ Boot transaction
+    }
+
+makeFields ''TreeGenGlobalCtx
+
+initTreeGenGlobalCtx :: Transaction h Addr -> TreeGenGlobalCtx h
+initTreeGenGlobalCtx boot = TreeGenGlobalCtx
+  { _treeGenGlobalCtxFreshHashSrc = 1
+  , _treeGenGlobalCtxInputPartiesUpperLimit = 1
+  , _treeGenGlobalCtxForkLikelihood = 0.1
+  , _treeGenGlobalCtxPruneLikelihood = 0.3
+  , _treeGenGlobalCtxMaxHeight = 50
+  , _treeGenGlobalCtxSharedTransactionLikelihood = 0.5
+  , _treeGenGlobalCtxBootTransaction = boot
+  }
+
+data TreeGenBranchCtx h = TreeGenBranchCtx
+    { _treeGenBranchCtxPrincipalBranch :: !Bool
+      -- ^ Is this the principal branch? The principal branch will never
+      --   be pruned.
+    , _treeGenBranchCtxBranchHeight    :: !Int
+      -- ^ Height of the tip of the branch (from the root).
+    , _treeGenBranchCtxCurrentUtxo     :: !(Utxo h Addr)
+      -- ^ The mapping of current addresses and their current account values.
+    , _treeGenBranchCtxTransactions    :: ![Value -> Transaction h Addr]
+      -- ^ Transactions to form this block. This is stored here because we
+      --   generate all transactions for the child block at the parent level,
+      --   in order to facilitate sharing.
+    }
+
+makeFields ''TreeGenBranchCtx
+
+-- | Tree generation monad.
+newtype TreeGen h a = TreeGen
+    { unTreeGen :: StateT (TreeGenGlobalCtx h) Gen a
+    } deriving (Functor, Applicative, Monad, MonadState (TreeGenGlobalCtx h))
+
+-- | Lift a 'Gen' action into the 'TreeGen' monad.
+liftGenTree :: Gen a -> TreeGen h a
+liftGenTree = TreeGen . lift
+
+genValidBlocktree :: Hash h Addr => PreTree h Gen ()
+genValidBlocktree = toPreTreeWith identity newTree
+
+toPreTreeWith
+    :: Hash h Addr
+    => (TreeGenGlobalCtx h -> TreeGenGlobalCtx h) -- ^ Modify the global settings
+    -> TreeGen h (NoFeeBlockTree h)
+    -> PreTree h Gen ()
+toPreTreeWith settings bg = DepIndep $ \(boot :: Transaction h Addr)-> do
+    ks <- evalStateT (unTreeGen bg) (settings (initTreeGenGlobalCtx boot))
+    return $ \fees ->
+      (OldestFirst (fmap (OldestFirst . dropWhile (== boot))
+                         (zipTreeFees ks fees)), ())
```

Comment:
This formatting is a tiny bit confusing to me; for multi-line tuples I prefer to write:

```haskell
x = 
    ( foo
    , bar
    )
```

to make it clearer what's going on.